feat: docs (#20)
cowprotocol/euler-integration-contracts: docs/01-overview.md

# Euler-CoW Protocol Integration: Wrappers Overview

## Introduction

> This repository contains **Euler-CoW Protocol integration contracts** that enable leveraged position management (opening, closing, growing, and shrinking) and collateral swaps through CoW Protocol settlements combined with [Ethereum Vault Connector (EVC)](https://evc.wtf/) operations. The contracts, using CoW's [Generalized Wrappers](https://docs.cow.fi/cow-protocol/concepts/order-types/wrappers) architecture, coordinate complex multi-step DeFi operations atomically around a CoW settlement operation. Through its

Reviewer: The link to generalized wrapper is broken, which makes sense since they are introduced in this repo as well. Since they're also audited, they should also be documented and they probably deserve their own page in the current docs structure.

Author: ok makes sense. Will try and draft something up.

[`CowWrapper.sol`](../../src/CowWrapper.sol) is a self-contained abstract base contract provided by the CoW DAO which should be used for all wrappers. In particular it ensures:

- **Solver Authentication**: Verifies that only authenticated CoW Protocol solvers can initiate settlements
- **Wrapper Chaining**: Allows multiple wrappers to be chained together, where each wrapper processes its own logic before delegating to the next wrapper or final settlement
- **Settlement Routing**: Routes the settlement call through the wrapper chain to the CoW Protocol settlement contract

### EVC Integration Base: CowEvcBaseWrapper

[`CowEvcBaseWrapper.sol`](../../src/CowEvcBaseWrapper.sol) extends `CowWrapper` with:

- **EVC Batch Coordination**: Manages batching of operations within the [EVC's atomic execution context](https://evc.wtf/docs/concepts/internals/batch)
- **Authorization Mechanisms**: Supports two authorization flows:
  - **EVC Permit Flow**: Users sign a [permit message](https://evc.wtf/docs/concepts/internals/permit/) for one-time authorization with the EVC
  - **Pre-Approved Hash Flow**: Users pre-approve operation hashes on-chain (useful for EIP-7702 wallets)

> - **Parameter Hashing**: Uses EIP-712 to securely hash parameters and verify user intent

Reviewer: Imho this is kind of weird, it's too specific to be in an overview.

```suggestion
```

- **Account Health Checks**: Leverages EVC's automatic account status checks at batch conclusion

## The Wrappers

### 1. CowEvcOpenPositionWrapper

> The [`CowEvcOpenPositionWrapper`](../../src/CowEvcOpenPositionWrapper.sol) opens or grows leveraged positions (long or short).

Reviewer: Broken link. Same comment for all instances of `src`.

```suggestion
The [`CowEvcOpenPositionWrapper`](../src/CowEvcOpenPositionWrapper.sol) opens or grows leveraged positions (long or short).
```

> The [`CowEvcOpenPositionWrapper`](../../src/CowEvcOpenPositionWrapper.sol) opens or grows leveraged positions (long or short).

Reviewer: Maybe stupid question but what does long or short mean here?
In the example below, I'd say the position is short in ETH. It can arguably be said to be long in USDC but anyway, I'd think of this mainly as a short.

I'm even considering if we should call the OpenPosition wrapper a Short wrapper. (But then it isn't nicely symmetric with the next one, so maybe not.)

Author: obviously I am not a trading expert. When I say "long x", it means that I am creating exposure to x such that as x price rises, I get money, and when "short x", I gain money when x price falls. Generally there is some "safe haven" collateral asset that is used , which is generally USD, which is why I suppose its kind of unusual to say you are "long USD" because if X's price falls, that doesn't necessarily mean that you gained money because USD went up.

> I'm even considering if we should call the OpenPosition wrapper a Short wrapper. (But then it isn't nicely symmetric with the next one, so maybe not.)

I suppose it would make sense here to give examples of how it works both "long" and "short". The open position wrapper can also be a long wrapper. Consider if the user started with ETH and then borrowed USD, its wierd to say you are "short USD" because USD is not a volatile asset, so instead we just say you are long ETH. So whatever asset is used as collateral is always "long" and the collateral used as debt is always "short", but that doesn't make it a "short wrapper" I would say.

**Result**: User's debt is repaid and remaining collateral is left in the account

See the [dedicated page](./03-close-position.md) on this wrapper.

### 3. CowEvcCollateralSwapWrapper

The [`CowEvcCollateralSwapWrapper`](../../src/CowEvcCollateralSwapWrapper.sol) swaps all or a portion of collateral between different Euler vaults while holding debt

**Example**: Using the position opened in the open position wrapper above as an example, the user swaps their full 6000 USDC collateral to ~0.06 BTC (1 BTC = $100000) collateral while maintaining their debt position.

**Result**: The user holds both 0.06 BTC as collateral assets against their ETH debt. User's collateral composition changes without closing the position

See the [dedicated page](./04-collateral-swap.md) on this wrapper.

## Authorization Flows

There are two different ways a user can authorize their order.

### Flow 1: EVC Permit (Off-Chain Signature)

> Users provide an EIP-712 EVC.permit signature of the data returned by `getPermitData(params)` authorizing a specific operation.
> 
> Additionally, the user signs a [EIP-712 order with CoW](https://docs.cow.fi/cow-protocol/reference/core/signing-schemes#eip-712) (in the case of the close position wrapper, its a EIP-1271 order with the same signing).
> 

Reviewer: Just stressing a bit why we want a specific signature type. 

```suggestion
Users provide an EIP-712 EVC.permit signature of the data returned by `getPermitData(params)` authorizing a specific operation.

Additionally, the user signs an order on CoW protocol. As this flow is intended to be off-chain first, the CoW order can be signed with either an [EIP-712 CoW signature](https://docs.cow.fi/cow-protocol/reference/core/signing-schemes#eip-712) or an equivalent mechanism for specific wrappers.

```

The full flow is:

1. User's browser creates the `params` for the wrapper and the trade they want to execute

> 2. If any approvals are required for the trade to succeed, the user needs to sign an on-chain transaction for these (see the specific section for the wrapper being executed),

Reviewer: I think the one-time part is true, but please double check that. It's just that it's weird otherwise, since this flow is intended to be off-chain. Naturally Euler can keep the approval limited, but ideally an unlimited approval should be an option.

```suggestion
2. If any approvals are required for the trade to succeed, the user needs to sign an one-time on-chain transaction for these (see the specific section for the wrapper being executed),
```

2. If any approvals are required for the trade to succeed, the user needs to sign an on-chain transaction for these (see the specific section for the wrapper being executed),
3. User's browser receives operation hash by calling the wrapper: `wrapper.getApprovalHash(params)`
4. User executes on-chain transaction `wrapper.setPreApprovedHash(hash, true)`
5. User's browser generates the corresponding CoW order to `params`
6. User executes on-chain transaction to the CoW settlement contract `settlement.setPreSignature(orderUid, true)`
7. Later, wrapper validates hash was pre-approved in contract storage
8. Hash is permanently consumed after use, cannot be replayed

**Advantages**: With [EIP-7702](https://eips.ethereum.org/EIPS/eip-7702) wallets and wallet batching through [`wallet_sendCalls`](https://docs.metamask.io/wallet/reference/json-rpc-methods/wallet_sendcalls), can batch all needed approvals. Can be gassless with [EIP-4337](https://eips.ethereum.org/EIPS/eip-4337). Works seamlessly with smart contract wallets.
**Disadvantages**: If the wallet is not capable of batching transactions, requires at least 3 extra on-chain transactions.

## Security Model

> ### Trusted Actor: Solvers

Reviewer: Nit, just because all the items in the list are unrelated to solvers.

```suggestion
### Trusted Components
```

Author: the purpose of this section is intended to explain why solvers can be trusted (both by onchain and offchain processes). The examples given are specific onchain safeguards designed to prevent solvers from having incentive to steal funds, and the mentioning of slashing is explaining that solvers are constrained by that offchain. Seems reasonable to me?

# Euler-CoW Protocol Integration: Wrappers Overview

## Introduction

This repository contains **Euler-CoW Protocol integration contracts** that enable leveraged position management (opening, closing, growing, and shrinking) and collateral swaps through CoW Protocol settlements combined with Ethereum Vault Connector (EVC) operations. The contracts act as "wrappers" that coordinate complex multi-step DeFi operations atomically. Through its
design, further wrappers could be added in the future to satisfy evolving use-cases.

## What Are Wrappers?

Wrappers are smart contracts that add custom logic around [CoW Protocol settlements](https://docs.cow.fi/cow-protocol/reference/contracts/core/settlement). They enable transactions that would otherwise be impossible in a single atomic operation by coordinating multiple steps:

1. **Pre-settlement operations** (e.g., enabling vaults, approving collateral)
2. **CoW Protocol settlement** (DEX aggregation and order matching, converting to Euler vault tokens)
3. **Post-settlement operations** (e.g., repaying debt, adding collateral to user's vault)

> All of these steps execute atomically within an EVC batch, allowing for flash-loan like functionality where tokens can be borrowed before the supporting collateral exists.

Reviewer 2: User doesn't necesarilly have context on what is EVC batch. Could you link to a relevant page in Euler's doc?

Author: linked in all doc files where relevant

## Architecture Overview

### Base Framework: CowWrapper

> `CowWrapper.sol` is a self-contained abstract base contract provided by the CoW DAO which should be used for all wrappers. In particular it ensures:

Reviewer 2: I think is nice to link to the file

Author: linked to the specific file

- **Solver Authentication**: Verifies that only authenticated CoW Protocol solvers can initiate settlements
- **Wrapper Chaining**: Allows multiple wrappers to be chained together, where each wrapper processes its own logic before delegating to the next wrapper or final settlement
- **Settlement Routing**: Routes the settlement call through the wrapper chain to the CoW Protocol settlement contract

### EVC Integration Base: CowEvcBaseWrapper

> `CowEvcBaseWrapper.sol` extends `CowWrapper` with:

Reviewer 2: Same

Author: linked to the specific file

- **EVC Batch Coordination**: Manages batching of operations within the EVC's atomic execution context
- **Authorization Mechanisms**: Supports two authorization flows:
  - **EVC Permit Flow**: Users sign a permit message for one-time authorization with the EVC
  - **Pre-Approved Hash Flow**: Users pre-approve operation hashes on-chain (useful for EIP-7702 wallets)
- **Parameter Hashing**: Uses EIP-712 to securely hash parameters and verify user intent
- **Account Health Checks**: Leverages EVC's automatic account status checks at batch conclusion

## The Wrappers

### 1. CowEvcOpenPositionWrapper

**Purpose**: Opens or grows leveraged positions (long or short)

**Flow**:
1. Wrapper validates user authorization (permit or pre-approved hash)

> 2. Enable collateral vault as collateral

Reviewer 2: nice to say, where is it enabled, or what it means. If relevant, link to Euler docs relevant page

Author: linked to relevant page

> 3. Enable borrow vault as controller

Reviewer 2: Same, for the unfamiliar this is not very clear. Please link to docs

Reviewer 2: https://docs.euler.finance/concepts/core/evc#vault-mediation

Author: linked to appropriate corresponding docs in EVC website

4. Deposit user's margin collateral as needed
5. Borrow assets against the collateral

> 6. Execute CoW settlement to swap borrowed assets → collateral assets

Reviewer 2: ubernit: personally would not mix, either schematic with arrows, or plain english, but this one feels like its mixing the two

Author: replaced with `to` in all cases across all files

> 7. All operations occur atomically within EVC batch

Reviewer 2: Ubernit:

```suggestion
7.  Operations 1 to 6 occur atomically within the [EVC batch](https://docs.euler.finance/concepts/core/evc/#batching)
```

Author: this section removed in favor of detailed doc for each wrapper.

> 8. EVC validates account is sufficiently collateralized (above minimum collateral ratio)

Reviewer 2: I wonder if we need so much detail given us an overview, maybe high-level is enough, and then we link to the details (there's a page for it)

Author: yep, removed in favor of the detail already provided in the specific wrapper doc

**Result**: User holds a leveraged position with borrowed assets converted to additional collateral

**Example**: User deposits 1000 USDC, borrows 5 ETH (when 1 ETH = $1000), swaps those 5 ETH back to $5000 USDC. Result: 6000 USDC collateral backing $5000 WETH debt (120% collateralization).

### 2. CowEvcClosePositionWrapper

**Purpose**: Closes leveraged positions (full or partial)

**Flow**:
1. Wrapper validates user authorization

> 2. Create or reuse an Inbox contract (temporary fund holder unique to each Euler subaccount for security)

Reviewer 2: introduction here of "Euler subaccount" without context

Author: added links for euler subaccount and similar concepts through all files.

3. Transfer collateral from subaccount to Inbox

> 4. Execute CoW settlement to swap collateral → debt repayment assets

Reviewer 2: Its hard to get what is "swap colleateral" or "debt replayment assets" given the porpouse above.

Author: this section has been removed in favor of the detailed doc in the specific wrapper md.

5. Repay debt on user's behalf using swapped assets
6. Return excess assets to user's account
7. All operations occur atomically within EVC batch

**Special Considerations**:

> - **Unlike other wrappers, uses an `Inbox` contract as the settlement receiver and sender. This means that this wrapper should be signed as a EIP-1271 order.**

Reviewer 2: >  this wrapper should be signed as a EIP-1271 order

I don't understand exactly what signing a wrapper means.

Author: moved the details of what this entails to the detailed wrapper doc

> - Inbox temporarily holds swapped funds and manages repayment

Reviewer 2: nit: temporary is not clear enough. You expect it to receive the funds from the swap, but release them within the same transaction, no?

Author: clarified that it holds funds during the wrapper execution

- Supports both full repayment (with CoW order kind `KIND_BUY`) and partial repayment (`KIND_SELL`)

> **Result**: User's debt is repaid and remaining collateral is returned

Reviewer 2: I would even add the result before the flow.
Also, if we add the result, would be very clarifying what's the input

Author: changed the order of result and example in all cases. Also clarified the input in all cases.

> **Example**: User closes a 5 ETH short position (ETH = $1000). Around 5000 USDC collateral is swapped to exactly the user's debt of 5 ETH. The debt is repaid, and remaining USDC ($1000) is returned to the owner's account.

Reviewer 2: I would prefer the example before the detailed flow, as it helps to picture better the flow

Reviewer 2: Why here the debt is $4000?

Author: Sorry, I didnt clarify here the starting case is same as where the open position wrapper left off. 5 ETH at $1000 is around $5000 of debt, so starting with $6000 of debt, you end up with $1000 left over. added some clarification.

### 3. CowEvcCollateralSwapWrapper

**Purpose**: Swaps all or a portion of collateral between different Euler vaults while holding debt

**Flow**:
1. Wrapper validates user authorization
2. Enable destination vault as new collateral

> 3. Transfer collateral from subaccount to main account (if using subaccount)

Reviewer 2: The reader doesn't need to know what is the main account or subaccounts

https://docs.euler.finance/concepts/advanced/sub-accounts?_highlight=accounts

Author: removed, in the detailed wrapper docs now. also added links throughout the repo for subaccounts with the evc

4. Execute CoW settlement to swap old collateral → new collateral
5. New collateral is automatically deposited in the destination vault
6. All operations occur atomically within EVC batch
7. EVC validates account is sufficiently collateralized after swap (above minimum collateral ratio)

**Result**: User's collateral composition changes without closing the position

**Example**: User swaps from 1000 USDC collateral to 1 WETH collateral while maintaining their debt position.

> ## How Wrappers Fit Into CoW Orders

Reviewer 2: I would think this belongs before the detailed wrapper flows.

Author: removed. The explanation is mostly the same for all three wrappers, so moving to docs repo doc on wrappers

> Each wrapper is referenced in the CoW order's `appData` along with its encoded parameters:

Reviewer 2: link to cow docs for appData

Author: ended up being removed entirely from this documentation, as the documentation is in the wrapper documentatino (separate PR in docs repo)
